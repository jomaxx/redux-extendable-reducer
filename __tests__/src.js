import { createStore } from 'redux';
import rootReducer, { extendReducer } from '../src';

const done = (state = [], action) => state.concat(action);
const doneAction = { type: 'DONE' };

const consoleWarn = console.warn;

beforeAll(() => {
  console.warn = jest.fn(consoleWarn);
});

beforeEach(() => {
  jest.resetAllMocks();
});

afterAll(() => {
  console.warn = consoleWarn;
});

it('extends reducer', () => {
  const store = createStore(rootReducer);
  store.dispatch(extendReducer({ done }));
  store.dispatch(doneAction);
  expect(store.getState().done).toMatchSnapshot();
});

it('buffers actions until reducer is extended', () => {
  const serverStore = createStore(rootReducer);
  serverStore.dispatch(extendReducer({ done }));
  const initialState = JSON.parse(JSON.stringify(serverStore.getState()));
  const clientStore = createStore(rootReducer, initialState);
  clientStore.dispatch(doneAction);
  clientStore.dispatch(extendReducer({ done }));
  expect(clientStore.getState().done).toMatchSnapshot();
});

it('throws error when reducer is not a function', () => {
  const store = createStore(rootReducer);
  expect(() => store.dispatch(extendReducer({ done: true }))).toThrowErrorMatchingSnapshot();
});

it('warns when attempting to replace reducer', () => {
  console.warn.mockImplementationOnce(() => {});
  const store = createStore(rootReducer);
  store.dispatch(extendReducer({ done }));
  store.dispatch(extendReducer({ done() {} }));
  expect(console.warn.mock.calls).toMatchSnapshot();
});
