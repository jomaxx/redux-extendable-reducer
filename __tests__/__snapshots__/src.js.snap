exports[`test buffers actions until reducer is extended 1`] = `
Array [
  Object {
    "meta": Object {
      "keys": Array [
        "done",
      ],
    },
    "payload": Object {},
    "type": "@@redux-extendable-reducer/EXTEND_REDUCER",
  },
  Object {
    "type": "@@redux/INIT",
  },
  Object {
    "type": "DONE",
  },
  Object {
    "meta": Object {
      "keys": Array [
        "done",
      ],
    },
    "payload": Object {
      "done": [Function],
    },
    "type": "@@redux-extendable-reducer/EXTEND_REDUCER",
  },
]
`;

exports[`test extends reducer 1`] = `
Array [
  Object {
    "meta": Object {
      "keys": Array [
        "done",
      ],
    },
    "payload": Object {
      "done": [Function],
    },
    "type": "@@redux-extendable-reducer/EXTEND_REDUCER",
  },
  Object {
    "type": "DONE",
  },
]
`;

exports[`test throws error when reducer is not a function 1`] = `"reducers.done is not a function"`;

exports[`test warns when attempting to replace reducer 1`] = `
Array [
  Array [
    "attempt to replace reducers.done was ignored.",
  ],
]
`;
